File: apps/ios/FlexTranslate/Sources/FlexTranslate/DialogueTurn.swift
import Foundation

/// Одна завершённая реплика в двустороннем диалоге.
///
/// Иммутабельна, кроме translatedText и translationReason: они переходят из nil/pending
/// в реальное значение, когда MT-воркер отработает. Любое обновление делает новую копию
/// через withTranslation() — на месте ничего не правим.
///
/// - id:                  стабильный id для ForEach в SwiftUI (строка UUID).
/// - monotonicTs:         монотонная метка времени (мс) в момент финализации — для порядка.
/// - spokenLanguage:      FlexLanguage, на котором говорил спикер.
/// - originalText:        настоящий вывод ASR (никогда не выдуманный).
/// - translatedText:      настоящий вывод MT, когда готов; nil пока pending или если перевод закрыт.
/// - translationReason:   честная причина, если перевод заблокирован; nil при успехе или
///                        пока перевод ещё не пробовали.
/// - translationLanguage: FlexLanguage, на котором выражен translatedText (язык собеседника
///                        на момент реплики).
/// - mtEngineUsed:        человекочитаемая метка движка, который дал translatedText.
///                        nil пока реплика в pending.
struct DialogueTurn: Identifiable, Equatable, Sendable {
    let id: String
    let monotonicTs: Int64
    let spokenLanguage: FlexLanguage
    let originalText: String
    let translatedText: String?
    let translationReason: String?
    let translationLanguage: FlexLanguage
    let mtEngineUsed: String?

    init(
        id: String = UUID().uuidString,
        monotonicTs: Int64,
        spokenLanguage: FlexLanguage,
        originalText: String,
        translatedText: String? = nil,
        translationReason: String? = nil,
        translationLanguage: FlexLanguage,
        mtEngineUsed: String? = nil
    ) {
        self.id = id
        self.monotonicTs = monotonicTs
        self.spokenLanguage = spokenLanguage
        self.originalText = originalText
        self.translatedText = translatedText
        self.translationReason = translationReason
        self.translationLanguage = translationLanguage
        self.mtEngineUsed = mtEngineUsed
    }

    /// true, пока MT-воркер ещё не разрешил перевод этой реплики.
    var translationPending: Bool {
        translatedText == nil && translationReason == nil
    }

    /// Новая копия реплики с применённым результатом перевода.
    /// Ненулевым должно быть ровно одно из text/reason (как в TranslationResult).
    func withTranslation(text: String?, reason: String?, engineLabel: String? = nil) -> DialogueTurn {
        DialogueTurn(
            id: id,
            monotonicTs: monotonicTs,
            spokenLanguage: spokenLanguage,
            originalText: originalText,
            translatedText: text,
            translationReason: reason,
            translationLanguage: translationLanguage,
            mtEngineUsed: engineLabel
        )
    }
}

File: apps/android/app/src/main/java/dev/flextranslate/ui/DialogueTurn.kt
package dev.flextranslate.ui

/**
 * Одна завершённая реплика в двустороннем диалоге.
 *
 * Неизменяемая, кроме [translatedText] и [translationReason] — они переходят из null/в ожидании
 * в реальное значение, когда MT-воркер отработает. Обновление состояния всегда даёт новую копию
 * через [withTranslation], мутировать на месте нельзя.
 *
 * @param id           стабильный id для ключей Compose (строка UUID).
 * @param monotonicTs  монотонная метка времени (мс) на момент завершения реплики — для сортировки.
 * @param spokenLanguage  [FlexLanguage], на котором говорил собеседник.
 * @param originalText    настоящий вывод ASR (ничего не выдумываем).
 * @param translatedText  реальный перевод, когда готов; null пока ждём или если заблокировано.
 * @param translationReason честная причина блокировки перевода; null если перевод удался или ещё
 *   не пытались (в ожидании).
 * @param translationLanguage [FlexLanguage], на котором выражен [translatedText]
 *   (язык собеседника на момент реплики).
 * @param mtEngineUsed человекочитаемое имя движка, выдавшего [translatedText], напр. "Gemini Flash"
 *   или id локальной модели. Null пока реплика в ожидании.
 */
data class DialogueTurn(
    val id: String,
    val monotonicTs: Long,
    val spokenLanguage: FlexLanguage,
    val originalText: String,
    val translatedText: String?,
    val translationReason: String?,
    val translationLanguage: FlexLanguage,
    val mtEngineUsed: String? = null,
) {
    /** True, пока MT-воркер ещё не закрыл перевод этой реплики. */
    val translationPending: Boolean
        get() = translatedText == null && translationReason == null

    /**
     * Новая копия реплики с проставленным результатом перевода.
     * Ровно одно из [text] и [reason] должно быть не-null (как в [TranslationResult]).
     * [engineLabel] — имя движка, выдавшего [text] (напр. "Gemini Flash" или id локальной модели);
     * null, если перевод заблокирован или упал.
     */
    fun withTranslation(text: String?, reason: String?, engineLabel: String? = null): DialogueTurn =
        copy(translatedText = text, translationReason = reason, mtEngineUsed = engineLabel)
}
